### PR TITLE
bug: WriteObject now supports empty streams.

### LINKED ISSUE #2709: Cannot upload empty objects using Client::WriteObject()
We are not formatting the Range-Header for the final chunk correctly in this case.

### PR DESCRIPTION
**Pull request by coryan**
Support writing empty streams via `storage::Client::WriteObject()`, the
code was not right at all. This changed the API (and ABI) for a number
of classes in the `storage::internal::` namespace, so we have to update
the API/ABI baseline.

Fixes #2709.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2714)

### CODE REVIEW

**Comment by coryan** [comment]
Here is the compatibility report in case anybody is interested.

[storage_client_ reference to current compatibility report.pdf](https://github.com/googleapis/google-cloud-cpp/files/3241904/storage_client_.reference.to.current.compatibility.report.pdf)

**Comment by coryan** [comment]
PTAL: Rebase and resolved conflicts.

File: google/cloud/storage/client.cc [hunk 1]
```diff
@@ -166,13 +161,16 @@ StatusOr<ObjectMetadata> Client::UploadStreamResumable(
...
-    if (gcount < buffer.size()) {
-      source_size = session->next_expected_byte() + gcount;
-    }
+    bool final_chunk = (gcount < buffer.size());
+    auto source_size = session->next_expected_byte() + gcount;
     buffer.resize(gcount);
 
     auto expected = session->next_expected_byte() + gcount - 1;
-    upload_response = session->UploadChunk(buffer, source_size);
+    if (!final_chunk) {
```
Comment (devbww) [inline google/cloud/storage/client.cc]: Nit: I'd invert this if/else.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc [hunk 2]
```diff
@@ -92,48 +92,65 @@ std::streamsize CurlResumableStreambuf::xsputn(char const* s,
...
 
-StatusOr<HttpResponse> CurlResumableStreambuf::Flush(bool final_chunk) {
+StatusOr<HttpResponse> CurlResumableStreambuf::FlushFinal() {
+  // Shorten the buffer to the actual used size.
+  auto actual_size = static_cast<std::size_t>(pptr() - pbase());
+  std::size_t upload_size = upload_session_->next_expected_byte() + actual_size;
+  current_ios_buffer_.resize(actual_size);
+  hash_validator_->Update(current_ios_buffer_);
+
+  StatusOr<ResumableUploadResponse> result;
```
Comment (devbww) [inline google/cloud/storage/internal/curl_resumable_streambuf.cc]: Combine definition and initialization?

File: google/cloud/storage/internal/curl_resumable_streambuf.cc [hunk 3]
```diff
@@ -92,48 +92,65 @@ std::streamsize CurlResumableStreambuf::xsputn(char const* s,
...
+  }
+  // Reset the iostream put area with valid pointers, but empty.
+  current_ios_buffer_.resize(1);
+  auto pbeg = &current_ios_buffer_[0];
+  setp(pbeg, pbeg);
+
+  upload_session_.reset();
+
+  // If `result.ok() == false` we never get to this point, so the last response
+  // was actually successful, represent that by a HTTP 200 status code.
```
Comment (devbww) [inline google/cloud/storage/internal/curl_resumable_streambuf.cc]: s/, represent/. Represent/

File: google/cloud/storage/internal/curl_resumable_streambuf.cc [hunk 4]
```diff
@@ -92,48 +92,65 @@ std::streamsize CurlResumableStreambuf::xsputn(char const* s,
...
-  }
+  auto chunk_count = actual_size / UploadChunkRequest::kChunkSizeQuantum;
+  auto chunk_size = chunk_count * UploadChunkRequest::kChunkSizeQuantum;
+  std::string not_sent(pbase() + chunk_size, pbase() + actual_size);
+  current_ios_buffer_.assign(pbase(), pbase() + chunk_size);
   hash_validator_->Update(current_ios_buffer_);
 
-  auto result = upload_session_->UploadChunk(current_ios_buffer_, upload_size);
+  StatusOr<ResumableUploadResponse> result;
+  result = upload_session_->UploadChunk(current_ios_buffer_);
```
Comment (devbww) [inline google/cloud/storage/internal/curl_resumable_streambuf.cc]: Combine defn/init?

File: google/cloud/storage/internal/curl_resumable_streambuf.cc [hunk 5]
```diff
@@ -92,48 +92,65 @@ std::streamsize CurlResumableStreambuf::xsputn(char const* s,
...
+StatusOr<HttpResponse> CurlResumableStreambuf::FlushFinal() {
+  // Shorten the buffer to the actual used size.
+  auto actual_size = static_cast<std::size_t>(pptr() - pbase());
+  std::size_t upload_size = upload_session_->next_expected_byte() + actual_size;
+  current_ios_buffer_.resize(actual_size);
+  hash_validator_->Update(current_ios_buffer_);
+
+  StatusOr<ResumableUploadResponse> result;
+  result = upload_session_->UploadFinalChunk(current_ios_buffer_, upload_size);
+  if (!result.ok()) {
```
Comment (devbww) [inline google/cloud/storage/internal/curl_resumable_streambuf.cc]: if (!result) ... ?

File: google/cloud/storage/internal/resumable_upload_session.h [hunk 6]
```diff
@@ -40,6 +40,17 @@ class ResumableUploadSession {
    * @return The result of uploading the chunk.
    */
   virtual StatusOr<ResumableUploadResponse> UploadChunk(
+      std::string const& buffer) = 0;
+
+  /**
+   * Uploads a chunk and returns the resulting response.
```
Comment (devbww) [inline google/cloud/storage/internal/resumable_upload_session.h]: Does `UploadFinalChunk()` deserve a different description to `UploadChunk()`?

File: google/cloud/storage/internal/retry_resumable_upload_session_test.cc [hunk 7]
```diff
@@ -337,12 +323,114 @@ TEST_F(RetryResumableUploadSessionTest, TooManyTransientOnReset) {
...
+      std::move(mock), LimitedErrorCountRetryPolicy(10).clone(),
+      ExponentialBackoffPolicy(10_ms, 160_ms, 2).clone());
+
+  StatusOr<ResumableUploadResponse> response =
+      session.UploadFinalChunk(payload, quantum);
+  EXPECT_FALSE(response.ok());
+  EXPECT_EQ(PermanentError().code(), response.status().code());
+}
+
+/// @test Verify that too many transients on UploadFinalChuk results in a
```
Comment (devbww) [inline google/cloud/storage/internal/retry_resumable_upload_session_test.cc]: s/Chuk/Chunk/

Comment (coryan) [reply google/cloud/storage/internal/retry_resumable_upload_session_test.cc]: Thanks, fixed.

File: google/cloud/storage/internal/resumable_upload_session.h (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/internal/resumable_upload_session.h]: Fixed.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: Done.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: Done.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: Done.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: Doh! Yes.

File: google/cloud/storage/client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/client.cc]: Done.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/internal/curl_resumable_streambuf.cc (hunk 5 above)
Comment (devbww) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: I don't see that one (but I do see a similar update below).

Comment (coryan) [reply google/cloud/storage/internal/curl_resumable_streambuf.cc]: Doh, I think I did it right this time.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**